Checksum aside

dariusk opened the pull request:
This fixes #28.

dariusk committed:
Aside about fixing integer underflow warnings, fix #28

rylev commented on src/multiboot-headers.md:
```diff
@@ -282,7 +282,21 @@ Again, we’ll use math to let the computer calculate the sum for us. We add up
-TODO: https://github.com/intermezzOS/book/issues/28
+> By the way...
+>
+> 0xe85250d6 is 3,897,708,758 in decimal, and the smallest number you can store as a 32-bit signed integer is -2,147,483,648, so we get an integer underflow. This doesn't matter for the checksum, but since underflows are usually a bug the assembler gives us a warning.
```
Can you change this to "The magic number, 0xe85250d6..." so that the reader doesn't have to associate a random hex number with something they read several sentences back?

rylev replied:
Sorry, I'm confused: 3,897,708,758 is a positive number. Why do we get underflow?

rylev replied:
I don't mean to pick on this line too much, but can we also assume that every programmer knows what signed and unsigned integers are? Ruby, for example, does not have such a concept.

dariusk replied:
Totally fair, I'll rework it.

dariusk replied:
Huh, looking into this is I just realized that the original description of what is happening here is mostly wrong. The more I looked into the original math (which I'd taken for granted), it didn't seem like an underflow or overflow would be happening at all. I ran the code for myself without the 0x100000000, and I wasn't able to coerce any warning or errors out of the nasm assembler, just a failed boot which looked like a bad checksum.

Turns out [the multiboot spec](http://nongnu.askapache.com/grub/phcoder/multiboot.pdf) says nothing at all about modulo operators. It says:

> The field `checksum` is a 32-bit unsigned value which, when added to the other magic fields (i.e. ‘magic’, ‘architecture’ and ‘header_length’), must have a 32-bit unsigned sum of zero.

In other words:

`checksum` + `magic_number` + `architecture` + `header_length` = 0

We could try and "solve for" `checksum` like so:

`checksum` =  -(`magic_number` + `architecture` + `header_length`)

But of course we can't represent -(`magic_number` + `architecture` + `header_length`) in a uint32, due to it being unsigned. So instead we come up with a number that "tricks" the ALU into giving us zero when we do the addition. Imagine for the sake of argument that `magic_number` + `architecture` + `header_length` somehow works out to be 0xFFFFFFFE. The number we'd add to that in order to make 0 would be 0x00000002. This is 0x100000000-0xFFFFFFFE, because 0x100000000 technically maps to 0 with all the wraparound math.

(On top of all that, the checksum isn't even a signed integer, which makes sense, there's no point in signing things at this low a level!.)

So I guess I got a bunch of really good practice explaining what a signed integer is in my local branch here. I'll rewrite this entire description to be accurate.

dariusk replied:
(The point is that neither modulo nor signed ints play into this. The only thing I really have to describe is overflowing and what that does.)

dariusk replied:
For the record, here's what I wrote on signed integers:

https://gist.github.com/dariusk/505305e584bd8a2006f5

I'm putting it here so that there's a record of it. We might need it later on in the book even if it isn't useful right here.

rylev commented on src/multiboot-headers.md:
```diff
@@ -282,7 +282,21 @@ Again, we’ll use math to let the computer calculate the sum for us. We add up
 from a big number. `dd` then puts that value into this spot in our file.
 
-TODO: https://github.com/intermezzOS/book/issues/28
+> By the way...
```
Ah sorry, I forgot to mention that this should be bold `**By the way...**`

dariusk replied:
Huh, strange. It's `**By the way...**` in my local source but not rendering that way here in the PR? I'll look into this...

dariusk replied:
Yeah, check this out, it's definitely [checked in as bold](https://raw.githubusercontent.com/dariusk/book/master/src/multiboot-headers.md). I'm going to assume it's a rendering issue with the diff here?

rylev replied:
Hmmm... I'm still not seeing it checked in that way.
<img width="245" alt="screen shot 2016-01-12 at 14 54 53" src="https://cloud.githubusercontent.com/assets/1327285/12265279/825f166c-b93c-11e5-9663-5280c3b9ed9a.png">

Could it be you're looking at another aside?

dariusk replied:
You're absolutely right.

dariusk committed:
Correct explanation for checksum

dariusk commented:
Okay, completely rewrote the explanation to correctly explain what's happening with the checksum here.

steveklabnik commented:
:+1: 

ugh this makes me so happy. This is exactly the kind of learning / documenting that never gets written down by people.

rylev commented on src/multiboot-headers.md:
```diff
@@ -282,7 +282,26 @@ Again, we’ll use math to let the computer calculate the sum for us. We add up
-TODO: https://github.com/intermezzOS/book/issues/28
+> **By the way...**
+>
+> You might wonder why we're subtracting these values from a big number. To answer this we can look at what [the multiboot spec](http://nongnu.askapache.com/grub/phcoder/multiboot.pdf) says about the checksum value in the header:
```
Can you repeat which "big number" you mean?

rylev commented on src/multiboot-headers.md:
```diff
@@ -282,7 +282,26 @@ Again, we’ll use math to let the computer calculate the sum for us. We add up
+>
+> You might wonder why we're subtracting these values from a big number. To answer this we can look at what [the multiboot spec](http://nongnu.askapache.com/grub/phcoder/multiboot.pdf) says about the checksum value in the header:
+>
+> > The field `checksum` is a 32-bit unsigned value which, when added to the other magic fields (i.e. `magic`, `architecture` and `header_length`), must have a 32-bit unsigned sum of zero.
```
Let's go ahead an assume an appendix on signed and unsigned numbers and link to it here. I volunteer to write it.

rylev commented:
Looking great! This is a tough one to explain as it requires some new concepts to wrap your head around. I like the way it's shaping up though!

dariusk committed:
Adding appendix stub for signed ints

dariusk committed:
Replacing vague ref with specific ref

dariusk commented:
I made a stub appendix and added a TODO issue #70 for it (@steveklabnik, can you label it? or give me label permissions?). Also addressed @rylev's comment about the "big number".

dariusk commented:
I did have one more question. I'm using entirely straight-quotes instead of smart-quotes in my passage. Is this something I should worry about in a PR? I'm using Vim to write this stuff and

rylev commented:
Awesome stuff. I went ahead and labeled the issue. Let's get this merged.

steveklabnik commented:
@dariusk i think that the answer is "dont worry about the curlies, i'll just fix them afterwards"

My vim is set up to insert them automatically, so it's easy for me, but not for everyone...
